File: en/codes/csharp/chapter_dynamic_programming/edit_distance.cs
/**
* File: edit_distance.cs
* Created Time: 2023-07-14
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_dynamic_programming;

public class edit_distance {
    /* Edit distance: Brute-force search */
    int EditDistanceDFS(string s, string t, int i, int j) {
        // If both s and t are empty, return 0
        if (i == 0 && j == 0)
            return 0;
        // If s is empty, return length of t
        if (i == 0)
            return j;
        // If t is empty, return length of s
        if (j == 0)
            return i;
        // If two characters are equal, skip both characters
        if (s[i - 1] == t[j - 1])
            return EditDistanceDFS(s, t, i - 1, j - 1);
        // Minimum edit steps = minimum edit steps of insert, delete, replace + 1
        int insert = EditDistanceDFS(s, t, i, j - 1);
        int delete = EditDistanceDFS(s, t, i - 1, j);
        int replace = EditDistanceDFS(s, t, i - 1, j - 1);
        // Return minimum edit steps
        return Math.Min(Math.Min(insert, delete), replace) + 1;
    }

    /* Edit distance: Memoization search */
    int EditDistanceDFSMem(string s, string t, int[][] mem, int i, int j) {
        // If both s and t are empty, return 0
        if (i == 0 && j == 0)
            return 0;
        // If s is empty, return length of t
        if (i == 0)
            return j;
        // If t is empty, return length of s
        if (j == 0)
            return i;
        // If there's a record, return it directly
        if (mem[i][j] != -1)
            return mem[i][j];
        // If two characters are equal, skip both characters
        if (s[i - 1] == t[j - 1])
            return EditDistanceDFSMem(s, t, mem, i - 1, j - 1);
        // Minimum edit steps = minimum edit steps of insert, delete, replace + 1
        int insert = EditDistanceDFSMem(s, t, mem, i, j - 1);
        int delete = EditDistanceDFSMem(s, t, mem, i - 1, j);
        int replace = EditDistanceDFSMem(s, t, mem, i - 1, j - 1);
        // Record and return minimum edit steps
        mem[i][j] = Math.Min(Math.Min(insert, delete), replace) + 1;
        return mem[i][j];
    }

    /* Edit distance: Dynamic programming */
    int EditDistanceDP(string s, string t) {
        int n = s.Length, m = t.Length;
        int[,] dp = new int[n + 1, m + 1];
        // State transition: first row and first column
        for (int i = 1; i <= n; i++) {
            dp[i, 0] = i;
        }
        for (int j = 1; j <= m; j++) {
            dp[0, j] = j;
        }
        // State transition: rest of the rows and columns
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (s[i - 1] == t[j - 1]) {
                    // If two characters are equal, skip both characters
                    dp[i, j] = dp[i - 1, j - 1];
                } else {
                    // Minimum edit steps = minimum edit steps of insert, delete, replace + 1
                    dp[i, j] = Math.Min(Math.Min(dp[i, j - 1], dp[i - 1, j]), dp[i - 1, j - 1]) + 1;
                }
            }
        }
        return dp[n, m];
    }

    /* Edit distance: Space-optimized dynamic programming */
    int EditDistanceDPComp(string s, string t) {
        int n = s.Length, m = t.Length;
        int[] dp = new int[m + 1];
        // State transition: first row
        for (int j = 1; j <= m; j++) {
            dp[j] = j;
        }
        // State transition: rest of the rows
        for (int i = 1; i <= n; i++) {
            // State transition: first column
            int leftup = dp[0]; // Temporarily store dp[i-1, j-1]
            dp[0] = i;
            // State transition: rest of the columns
            for (int j = 1; j <= m; j++) {
                int temp = dp[j];
                if (s[i - 1] == t[j - 1]) {
                    // If two characters are equal, skip both characters
                    dp[j] = leftup;
                } else {
                    // Minimum edit steps = minimum edit steps of insert, delete, replace + 1
                    dp[j] = Math.Min(Math.Min(dp[j - 1], dp[j]), leftup) + 1;
                }
                leftup = temp; // Update for next round's dp[i-1, j-1]
            }
        }
        return dp[m];
    }

    [Test]
    public void Test() {
        string s = "bag";
        string t = "pack";
        int n = s.Length, m = t.Length;

        // Brute-force search
        int res = EditDistanceDFS(s, t, n, m);
        Console.WriteLine("Change " + s + " to " + t + " requires a minimum of " + res + " edits");

        // Memoization search
        int[][] mem = new int[n + 1][];
        for (int i = 0; i <= n; i++) {
            mem[i] = new int[m + 1];
            Array.Fill(mem[i], -1);
        }

        res = EditDistanceDFSMem(s, t, mem, n, m);
        Console.WriteLine("Change " + s + " to " + t + " requires a minimum of " + res + " edits");

        // Dynamic programming
        res = EditDistanceDP(s, t);
        Console.WriteLine("Change " + s + " to " + t + " requires a minimum of " + res + " edits");

        // Space-optimized dynamic programming
        res = EditDistanceDPComp(s, t);
        Console.WriteLine("Change " + s + " to " + t + " requires a minimum of " + res + " edits");
    }
}

File: ja/codes/go/chapter_dynamic_programming/edit_distance.go
// File: edit_distance.go
// Created Time: 2023-07-23
// Author: Reanon (793584285@qq.com)

package chapter_dynamic_programming

/* 編集距離：総当たり探索 */
func editDistanceDFS(s string, t string, i int, j int) int {
	// s と t がともに空なら 0 を返す
	if i == 0 && j == 0 {
		return 0
	}
	// s が空なら t の長さを返す
	if i == 0 {
		return j
	}
	// t が空なら s の長さを返す
	if j == 0 {
		return i
	}
	// 2 つの文字が等しければ、その 2 文字をそのままスキップする
	if s[i-1] == t[j-1] {
		return editDistanceDFS(s, t, i-1, j-1)
	}
	// 最小編集回数 = 挿入・削除・置換の 3 操作における最小編集回数 + 1
	insert := editDistanceDFS(s, t, i, j-1)
	deleted := editDistanceDFS(s, t, i-1, j)
	replace := editDistanceDFS(s, t, i-1, j-1)
	// 最小編集回数を返す
	return MinInt(MinInt(insert, deleted), replace) + 1
}

/* 編集距離：メモ化探索 */
func editDistanceDFSMem(s string, t string, mem [][]int, i int, j int) int {
	// s と t がともに空なら 0 を返す
	if i == 0 && j == 0 {
		return 0
	}
	// s が空なら t の長さを返す
	if i == 0 {
		return j
	}
	// t が空なら s の長さを返す
	if j == 0 {
		return i
	}
	// 記録済みなら、それをそのまま返す
	if mem[i][j] != -1 {
		return mem[i][j]
	}
	// 2 つの文字が等しければ、その 2 文字をそのままスキップする
	if s[i-1] == t[j-1] {
		return editDistanceDFSMem(s, t, mem, i-1, j-1)
	}
	// 最小編集回数 = 挿入・削除・置換の 3 操作における最小編集回数 + 1
	insert := editDistanceDFSMem(s, t, mem, i, j-1)
	deleted := editDistanceDFSMem(s, t, mem, i-1, j)
	replace := editDistanceDFSMem(s, t, mem, i-1, j-1)
	// 最小編集回数を記録して返す
	mem[i][j] = MinInt(MinInt(insert, deleted), replace) + 1
	return mem[i][j]
}

/* 編集距離：動的計画法 */
func editDistanceDP(s string, t string) int {
	n := len(s)
	m := len(t)
	dp := make([][]int, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = make([]int, m+1)
	}
	// 状態遷移：先頭行と先頭列
	for i := 1; i <= n; i++ {
		dp[i][0] = i
	}
	for j := 1; j <= m; j++ {
		dp[0][j] = j
	}
	// 状態遷移: 残りの行と列
	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			if s[i-1] == t[j-1] {
				// 2 つの文字が等しければ、その 2 文字をそのままスキップする
				dp[i][j] = dp[i-1][j-1]
			} else {
				// 最小編集回数 = 挿入・削除・置換の 3 操作における最小編集回数 + 1
				dp[i][j] = MinInt(MinInt(dp[i][j-1], dp[i-1][j]), dp[i-1][j-1]) + 1
			}
		}
	}
	return dp[n][m]
}

/* 編集距離：空間最適化した動的計画法 */
func editDistanceDPComp(s string, t string) int {
	n := len(s)
	m := len(t)
	dp := make([]int, m+1)
	// 状態遷移：先頭行
	for j := 1; j <= m; j++ {
		dp[j] = j
	}
	// 状態遷移：残りの行
	for i := 1; i <= n; i++ {
		// 状態遷移：先頭列
		leftUp := dp[0] // dp[i-1, j-1] を一時保存する
		dp[0] = i
		// 状態遷移：残りの列
		for j := 1; j <= m; j++ {
			temp := dp[j]
			if s[i-1] == t[j-1] {
				// 2 つの文字が等しければ、その 2 文字をそのままスキップする
				dp[j] = leftUp
			} else {
				// 最小編集回数 = 挿入・削除・置換の 3 操作における最小編集回数 + 1
				dp[j] = MinInt(MinInt(dp[j-1], dp[j]), leftUp) + 1
			}
			leftUp = temp // 次の反復の dp[i-1, j-1] に更新する
		}
	}
	return dp[m]
}

func MinInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}
